### PR TITLE
Show the app version in the settings view

### PR DESCRIPTION
**Pull request by eriklarko**
@anderklander Should I put the version stuff in `app/lib/platform.js`?

---
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/49)

### CODE REVIEW

File: app/components/Settings.js [hunk 1]
```diff
@@ -100,10 +104,28 @@ export default class Settings extends Component {
...
+  }
+
+  _formattedVersion() {
+    // the version in package.json has to be semver, but we use a YEAR.release-channel
+    // version scheme. in package.json we thus have to write YEAR.release.X-channel and
+    // this function is responsible for removing .X part.
+    const version = remote.app.getVersion();
+    return version
+      .replace('.0-', '-')  // remove the .0 in 2018.1.0-beta9
+      .replace(/\.0$/, ''); // remove the .0 in 2018.1.0
```
Comment (faern) [inline app/components/Settings.js]: Might want to have `\.` in the first regex as well, to escape the dot.

**Comment by eriklarko** [comment]
*[app/components/Settings.js, line 126 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4BhgR0iJ-nwAm3H2RN-r1-126:-L4FUDqfKEv1hA-dsnNP:b-h23ucb) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d1229cd66a04566422671f91a3a2d894cbe01e71/app/components/Settings.js#L126)):*
When the first argument is a string it is not as a regex, https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace

```
> 'foo.0'.replace('.', '-bar-')
'foo-bar-0'
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by eriklarko** [comment]
*[app/components/Settings.js, line 126 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4BhgR0iJ-nwAm3H2RN-r1-126:-L4FUYGZnJejba4UsZi6:b-wm7ypd) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d1229cd66a04566422671f91a3a2d894cbe01e71/app/components/Settings.js#L126)):*
<details><summary><i>Previously, eriklarko (Erik Larkö) wrote…</i></summary><blockquote>

When the first argument is a string it is not as a regex, https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace

```
> 'foo.0'.replace('.', '-bar-')
'foo-bar-0'
```
</blockquote></details>

...it is not interpreted as a regex *

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by faern** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[app/components/Settings.js, line 126 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4BhgR0iJ-nwAm3H2RN-r1-126:-L4FVb6k1hzrwVMicLkA:b62o606) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d1229cd66a04566422671f91a3a2d894cbe01e71/app/components/Settings.js#L126)):*
<details><summary><i>Previously, eriklarko (Erik Larkö) wrote…</i></summary><blockquote>

...it is not interpreted as a regex *
</blockquote></details>

Wow... Ok.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by eriklarko** [comment]
This is how it looks 

The scroll is broken however, so it's hard to quit the app with this patch :) I'll look into it a bit

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-:-L4FYiEreNmVBKZIGBx5:b-k7rpjm)*

**Comment by faern** [comment]
Hmm. Does it do anything if you click it? It looks just like the other items here, which all are buttons. It should not be shown in the same way as "Settings" is shown at the top?

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-:-L4FivrtstWAoySR7Dox:bn3k4nq)*

**Comment by pronebird** [comment]
What seems to be broken?

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-:-L4FjzhVaSg-SkCvJApO:bu7jgus)*

**Comment by eriklarko** [comment]
I can only reference the designs in the trello card https://trello.com/c/CakuMPYh/11-display-client-version-in-gui. There are some things there that this PR does not address, but it is a start and those things can be added later

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-:-L4FpDTY6fIONiGMk1-T:bx85wce)*

**Comment by pronebird** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[app/components/SettingsStyles.js, line 26 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4FqoGy8SoNyzBjCIrh:-L4FqoGy8SoNyzBjCIri:blq21m7) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bdce56631eaaf422a2278c3b2890f19eadbcf336/app/components/SettingsStyles.js#L26)):*
> ```JavaScript
>     justifyContent: 'space-between',
>     height: '100%',
>     overflow: 'visible',
> ```

`height: 100%` and `overflow: visible` are mutually exclusive in a way. Removing both should do the trick. The scrollbar will take care of the rest. Let them flow naturally.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by pronebird** [comment]
Reviewed 1 of 1 files at r2.
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/components/Settings.js, line 123 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4FsVBE442UlNIFWpwz:-L4FsVBE442UlNIFWpx-:brnennd) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bdce56631eaaf422a2278c3b2890f19eadbcf336/app/components/Settings.js#L123)):*
> ```JavaScript
>     // version scheme. in package.json we thus have to write YEAR.release.X-channel and
>     // this function is responsible for removing .X part.
>     const version = remote.app.getVersion();
> ```

In order to exclude the Electron from that formula we can add an "app version" prop to `Settings`, then do the mapping in `mapStateToProps`, for now a simple merge would work just fine: 

```
const mapStateToProps(state) => ({ ...state, appVersion: myGetAppVersion() })
```

It's not necessarily that we'll need to have a platform-function for that, we could actually read the version from `package.json`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by pronebird** [comment]
Reviewed 1 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by pronebird** [comment]
Hmm why not to add the link to the download page? Adding a new constant to `config.json` and calling `openLink` should be trivial. Do you need some help with this?

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-:-L4FtkMmWNnZ4UwmJA9i:bnns5gs)*

**Comment by eriklarko** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/components/Settings.js, line 123 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4FsVBE442UlNIFWpwz:-L4G6i_0NdujOU2-Cx4x:b7i0xzl) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bdce56631eaaf422a2278c3b2890f19eadbcf336/app/components/Settings.js#L123)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

In order to exclude the Electron from that formula we can add an "app version" prop to `Settings`, then do the mapping in `mapStateToProps`, for now a simple merge would work just fine: 

```
const mapStateToProps(state) => ({ ...state, appVersion: myGetAppVersion() })
```

It's not necessarily that we'll need to have a platform-function for that, we could actually read the version from `package.json`.
</blockquote></details>

Feels like we're losing too much cohesion. But I guess it does fit better into the smart/dump components thing... Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by eriklarko** [comment]
*[app/components/SettingsStyles.js, line 26 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4FqoGy8SoNyzBjCIrh:-L4G7nVvgy0yhQyfz-a5:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bdce56631eaaf422a2278c3b2890f19eadbcf336/app/components/SettingsStyles.js#L26)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

`height: 100%` and `overflow: visible` are mutually exclusive in a way. Removing both should do the trick. The scrollbar will take care of the rest. Let them flow naturally.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by pronebird** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/components/Settings.js, line 123 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4FsVBE442UlNIFWpwz:-L4G83wa-1FlXvXf0fn6:binmfl1) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bdce56631eaaf422a2278c3b2890f19eadbcf336/app/components/Settings.js#L123)):*
<details><summary><i>Previously, eriklarko (Erik Larkö) wrote…</i></summary><blockquote>

Feels like we're losing too much cohesion. But I guess it does fit better into the smart/dump components thing... Done
</blockquote></details>

I agree with you on this. 

I like that we can offload some of the logic to containers but I must admit that it's impossible to have any reasonable lifecycle in there so they are more like "reducers", however we can expose props returning Promises etc and do the rest of logic in the Component.

For example I'd like to move the user location fetch into the `Connect` so I can do it at more appropriate times. 

However that would complicate testing if we had the real IPC running there. 

As a solution to that problem I could simply expose the prop and hook it up from the container which would do the `backend.fetchLocation()` and return a `Promise` to the component. Then in testing I'd simply mock that call without a need to carry the IPC with me or even have a mock of it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by pronebird** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/components/Settings.js, line 123 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4FsVBE442UlNIFWpwz:-L4GAgW0vsu8MGVdEfkW:b-l1vc3w) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bdce56631eaaf422a2278c3b2890f19eadbcf336/app/components/Settings.js#L123)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

I agree with you on this. 

I like that we can offload some of the logic to containers but I must admit that it's impossible to have any reasonable lifecycle in there so they are more like "reducers", however we can expose props returning Promises etc and do the rest of logic in the Component.

For example I'd like to move the user location fetch into the `Connect` so I can do it at more appropriate times. 

However that would complicate testing if we had the real IPC running there. 

As a solution to that problem I could simply expose the prop and hook it up from the container which would do the `backend.fetchLocation()` and return a `Promise` to the component. Then in testing I'd simply mock that call without a need to carry the IPC with me or even have a mock of it.
</blockquote></details>

The last paragraph implied the prop that receives the function that returns a Promise.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by anderklander** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[app/containers/SettingsPage.js, line 6 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4GB0L1C_VSv3FTt527:-L4GC35k4KsqlLRJnijD:ba2mvzm) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bef515f47c4030aa89734023e8c41d53313acbdb/app/containers/SettingsPage.js#L6)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

Just so we don't have to deal with platform dependencies here. Wouldn't it suffice to `import { version } from '../../package.json'`?
</blockquote></details>

Agree on the above, or if not possible it would be great to move the electron stuff to platform for getting the version!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by eriklarko** [comment]
*[app/components/Settings.js, line 123 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4FsVBE442UlNIFWpwz:-L4GELtahaI3N2ZxY-Ma:b-aesh6m) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bdce56631eaaf422a2278c3b2890f19eadbcf336/app/components/Settings.js#L123)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

The last paragraph implied the prop that receives the function that returns a Promise.
</blockquote></details>

Not sure how to proceed now, but the version is passed as prop and read in the connected component

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by eriklarko** [comment]
Done.

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/containers/SettingsPage.js, line 6 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4GB0L1C_VSv3FTt527:-L4GFMKMTLautbxhM_HI:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bef515f47c4030aa89734023e8c41d53313acbdb/app/containers/SettingsPage.js#L6)):*
<details><summary><i>Previously, anderklander wrote…</i></summary><blockquote>

Agree on the above, or if not possible it would be great to move the electron stuff to platform for getting the version!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-:-L4GEJbjWIJXZKLVmPAT:b-896fix)*

**Comment by pronebird** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/components/Settings.js, line 123 at r2](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/49#-L4FsVBE442UlNIFWpwz:-L4GFQlWCHvzj7awOPGL:bxmx79a) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bdce56631eaaf422a2278c3b2890f19eadbcf336/app/components/Settings.js#L123)):*
<details><summary><i>Previously, eriklarko (Erik Larkö) wrote…</i></summary><blockquote>

Not sure how to proceed now, but the version is passed as prop and read in the connected component
</blockquote></details>

No no I was just giving an example about smart/dumb components. All good here!

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/49)*

**Comment by pronebird** [comment]
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/49)*

**Comment by eriklarko** [comment]
*[app/components/Settings.js, line 123 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-L4FsVBE442UlNIFWpwz:-L4GGat7H-0Gp3w4ehlT:b35i5ge) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bdce56631eaaf422a2278c3b2890f19eadbcf336/app/components/Settings.js#L123)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

No no I was just giving an example about smart/dumb components. All good here!
</blockquote></details>

Ah, cool :) Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by anderklander** [comment]
:lgtm:

---

Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-:-L4GHHHlXv6dHe0DQNoi:bnfp4nl)*

**Comment by pronebird** [comment]
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49)*

**Comment by pronebird** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/49#-:-L4GJZ59FXMQle8dDoVN:bnfp4nl)*

**Pull request merged**